bitcoin/src/bip32.rs

```diff
@@ -112,40 +112,96 @@ pub struct Xpub {
 #[cfg(feature = "serde")]
 crate::serde_utils::serde_string_impl!(Xpub, "a BIP-32 extended public key");
 
+/// A non-hardened child number.
+///
+/// # Note
+///
+/// Key index, within [0, 2^31 - 1].
+#[derive(Copy, Clone, PartialEq, Eq, Debug, PartialOrd, Ord, Hash)]
+pub struct NormalChildNumber(u32);
+
+impl NormalChildNumber {
+    /// Creates a new [`NormalChildNumber`] from an index.
+    pub fn new(index: u32) -> Self { NormalChildNumber(index) }
+
+    /// Returns the index of the child number.
+    pub fn index(&self) -> u32 { self.0 }
+}
+
+/// A hardened child number
+///
+/// # Note
+///
+/// Key index, within [0, 2^31 - 1].
+#[derive(Copy, Clone, PartialEq, Eq, Debug, PartialOrd, Ord, Hash)]
+pub struct HardenedChildNumber(u32);
+
+impl HardenedChildNumber {
+    /// Creates a new [`HardenedChildNumber`] from an index.
+    pub fn new(index: u32) -> Self { HardenedChildNumber(index) }
+
+    /// Returns the index of the child number.
+    pub fn index(&self) -> u32 { self.0 }
+}
+
+impl From<NormalChildNumber> for ChildNumber {
+    fn from(child: NormalChildNumber) -> Self { ChildNumber::Normal(child) }
+}
+
+impl TryFrom<ChildNumber> for NormalChildNumber {
+    type Error = Error;
```

Comment:
I'd prefer returning a more specific error but this can be done in a followup PR.